Make loan duration a per-library per-collection setting and make sure it's always corectly described.

leonardr opened the pull request:
This branch makes sure that all of our CollectionAPI subclasses (except the ones that offer unlimited-length loans) have one (ebooks) or two (ebooks+audiobooks) settings relating to loan durations.

For Overdrive, Axis, and Bibliotheca, these settings are informative only. They reflect the _default_ or _assumed average_ lengths of loans but the circulation manager has no power over the lengths of loans created through these services.

For ODL and RBdigital, these settings are active. When a loan is created, the duration of the loan is taken from this setting.

All this code existed before, but ODL and RBdigital were handled inconsistently, the settings were per-collection instead of per-library per-collection, and the descriptions of the settings didn't make it clear when these settings were informative and when they were active.

aslagle commented on api/circulation.py:
```diff
@@ -865,13 +865,46 @@ def sync_bookshelf(self, patron, pin):
+        "key" : Collection.EBOOK_LOAN_DURATION_KEY, 
+        "label": _("Ebook Loan Duration (in Days)"),
+        "default": Collection.STANDARD_DEFAULT_LOAN_PERIOD,
+        "description": _("When a patron uses SimplyE to borrow an audio book from this collection, SimplyE will ask for a loan that lasts this number of days. This must be equal to or less than the maximum loan duration negotiated with the distributor.")
```
This description should say "ebook" not "audio"

aslagle commented on api/odl.py:
```diff
@@ -100,17 +100,22 @@ class ODLWithConsolidatedCopiesAPI(BaseCirculationAPI):
+        BaseCirculationAPI.EBOOK_LOAN_DURATION_SETTING
+    )
+    my_ebook_loan_duration_setting.update(
+        description = _("When a patron uses SimplyE to borrow an ebook from this ODL server, SimplyE will ask for a loan that lasts this number of days. This must be equal to or less than the maximum loan duration negotiated with the distributor.")
```
I don't think it's necessary to replace this description.

leonardr commented:
Assuming the tests pass I'm happy with these changes. I changed core quite a bit and I removed assertions from the controller tests to reflect the fact that these configuration settings are no longer directly associated with the collection.

aslagle commented on api/odl.py:
```diff
@@ -158,7 +156,7 @@ def __init__(self, _db, collection):
         self.password = collection.external_integration.password
         self.consolidated_loan_url = collection.external_integration.setting(self.CONSOLIDATED_LOAN_URL_KEY).value
-        self.default_loan_period = collection.external_integration.setting(BaseCirculationAPI.DEFAULT_LOAN_PERIOD).int_value or 21
+        self.default_loan_period = collection.external_integration.setting(Collection.EBOOK_LOAN_DURATION_KEY).int_value or Collection.STANDARD_DEFAULT_LOAN_PERIOD
```
doesn't this need to change to be per-library?

aslagle commented on api/oneclick.py:
```diff
@@ -71,24 +71,29 @@ class OneClickAPI(BaseOneClickAPI, BaseCirculationAPI):
+        BaseCirculationAPI.EBOOK_LOAN_DURATION_SETTING
+    )
+    my_ebook_setting.update(
+        description = _("When a patron uses SimplyE to borrow an RBdigital ebook, SimplyE will ask for a loan that lasts this number of days. This must be equal to or less than the maximum loan duration negotiated with RBdigital."),
```
I'm not sure these descriptions need to be replaced either - they're the same as the default except with 'RBDigital' in them.

aslagle commented on api/oneclick.py:
```diff
@@ -103,11 +108,11 @@ def __init__(self, *args, **kwargs):
         )
         self.ebook_duration_days = (
-            integration.setting(self.EBOOK_LOAN_DURATION).int_value 
+            integration.setting(Collection.EBOOK_LOAN_DURATION_KEY).int_value 
```
these also don't use libraries.

aslagle commented on tests/admin/test_controller.py:
```diff
@@ -1969,14 +1957,6 @@ def test_collections_post_edit(self):
 
-        setting = collection.external_integration.setting("default_loan_period")
-        eq_("default_loan_period", setting.key)
-        eq_("14", setting.value)
```
This could be changed to check that the library + integration setting was set.

leonardr replied:
Can you tell me what the setup code would look like? I tried passing in a value with the library short name and nothing happened (which might either indicate a bug or that I did it wrong).

aslagle replied:
I think you have it right, but this is testing with an Overdrive collection and Overdrive doesn't have that setting.

aslagle replied:
Actually, I looked and the admin interface code for collections doesn't handle library settings yet - this is the first time we've had them for collections.

aslagle replied:
The JS part should be fine as is, but SettingsController.collections will need to be updated.

leonardr commented:
We need to add some code to the admin interface to handle setting library+collection-specific configuration information items. My brain can't handle this but Amy has agreed to tackle it tomorrow.

Here's the test I was working on when I discovered that the existing controller wasn't setting the configuration settings:

```
    def test_collections_post_edit_library_specific_configuration(self):
        # The collection exists.
        collection = self._collection(
            name="Collection 1",
            protocol=ExternalIntegration.RB_DIGITAL
        )

        l1, ignore = create(
            self._db, Library, name="Library 1", short_name="L1",
        )

        with self.app.test_request_context("/", method="POST"):
            flask.request.form = MultiDict([
                ("id", collection.id),
                ("name", "Collection 1"),
                ("protocol", ExternalIntegration.RB_DIGITAL),
                ("external_account_id", "1234"),
                ("username", "user2"),
                ("password", "password"),
                ("url", "http://rb/"),
                ("libraries", json.dumps([
                    {
                        "short_name": "L1", 
                        "ebook_loan_duration": "14",
                        "audio_loan_duration": "12"
                    }
                ])
                ),
            ])
            response = self.manager.admin_settings_controller.collections()
            eq_(response.status_code, 200)

        # Additional settings were set on the collection+library.
        set_trace()

        # Remove the connection between collection and library.
        with self.app.test_request_context("/", method="POST"):
            flask.request.form = MultiDict([
                ("id", collection.id),
                ("name", "Collection 1"),
                ("protocol", ExternalIntegration.OVERDRIVE),
                ("external_account_id", "1234"),
                ("username", "user2"),
                ("password", "password"),
                ("website_id", "1234"),
                ("ils_name", "the_ils"),
                ("libraries", json.dumps([])),
            ])
            response = self.manager.admin_settings_controller.collections()
            eq_(response.status_code, 200)

        eq_(collection.id, int(response.response[0]))

        # The settings associated with the collection+library were removed
        # when the connection between collection and library was deleted.

```
